### PR TITLE
NXP JN516x Platform

### PR DESCRIPTION
**Pull request by simonduq**
The JN516x series ([web page](http://www.nxp.com/techzones/wireless-connectivity/jn51xx/jn516x.html), [datasheet](http://www.nxp.com/documents/data_sheet/JN516X.pdf)) is a range of ultra low power, high performance wireless microcontrollers from NXP. They feature an enhanced 32-bit RISC processor (256kB/32kB/4kB Flash/RAM/EEPROM for JN5168), and also include a 2.4GHz IEEE802.15.4-compliant transceiver.

This pull request is a full port to JN516x, including support for various boards, timers, crypto engine, sensors. The port is documented (see the [platform readme file](https://github.com/simonduq/contiki/blob/jn516x-port/platform/jn516x/README.md)) and includes a new regression test for Travis. It is usable (compilation and tools) from Linux or Windows. It was tested thoroughly in a 25-node network supporting request-response traffic over RPL, where the application achieved >99.9% end-to-end packet delivery ratio over different MAC layers.

We have two long-term maintainers:
* Theo Van Daele, NXP, theo.van.daele@nxp.com, github user: [TeVeDe](https://github.com/TeVeDe)
* Simon Duquennoy, SICS, simonduq@sics.se, github user: [simonduq](https://github.com/simonduq)

Other contributors:
* Beshr Al Nahas, SICS (now Chalmers University), beshr@chalmers.se, github user: [beshrns](https://github.com/beshrns)
* Atis Elsts, SICS, atis.elsts@sics.se, github user: [atiselsts](https://github.com/atiselsts)

Additional long-term contact:
* Hugh Maaskant, NXP, hugh.maaskant@nxp.com, github user: [hugh-maaskant](https://github.com/hugh-maaskant)

### CODE REVIEW

File: examples/jn516x/rpl/border-router/border-router.c [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+/* While waiting for the prefix to be sent through the SLIP connection, the future
+ * border router can join an existing DAG as a parent or child, or acquire a default
+ * router that will later take precedence over the SLIP fallback interface.
+ * Prevent that by turning the radio off until we are initialized as a DAG root.
+ */
+  prefix_set = 0;
+
+  PROCESS_PAUSE();
+
+  PRINTF("RPL-Border router started\n");
```
Comment (alignan) [inline examples/jn516x/rpl/border-router/border-router.c]: I haven't read the full PR yet, but why is there another border router other than the one at examples/ipv6?

**Comment by adamdunkels** [comment]
Took a quick look and this looks very good overall, nice work on this!

Out of curiosity: is this to be considered an official NXP Contiki port? I see that there are a few NXP copyright headers in there. And if so, what is the relation to JIP?

**Comment by simonduq** [comment]
Thanks! There is significant contribution on the port yes.
@hugh-maaskant, @TeVeDe, I'll let you answer on whether this is an official NXP port and on the relation to JIP :)

File: examples/jn516x/rpl/border-router/border-router.c (hunk 1 above)
Comment (simonduq) [inline examples/jn516x/rpl/border-router/border-router.c]: Mostly the same, with two exceptions:
* `#include "dev/uart1.h"` replaced by `#include "dev/uart0.h"` in `slip-bridge.c` :p Here we would really need to try and make Contiki's border router example really platform-independent
* a number of functional changes: no web server and a convenient shared `common-conf.h` file used by the BR, the basic node, and different CoAP nodes for configuration

**Comment by mgielda** [comment]
We @antmicro have also been using this port in a project thanks to the kindness of @simonduq for some time, and would be very happy to see this PR merged.

**Comment by fros4943** [comment]
@simonduq Without actually having reviewed the PR yet, I just have to say that I really appreciate the presentation Simon! This is, IMO, a template for future new platform pull requests! :smile: I especially like that you included long-term maintainers.

**Comment by alignan** [comment]
@fros4943 @simonduq definitely! I will use this a reference to better document our platforms as well :-)

**Comment by simonduq** [comment]
@alignan well I have to acknowledge our readme file is heavily inspired by your latest (Re-Mote) ;)

**Comment by hugh-maaskant** [comment]
@adamdunkels, thanks for your comments.  I do not know what “official NXP port” means, but this work was largely done by SICS (@simonduq, @beshrns ) in the context of an ongoing collaboration to add support for TSCH and -parts of- 6TiSCH as well as adding the JN516x series to the list of supported devices in Contiki.  Obviously NXP is very happy to expose the Contiki community to such a great device family ;-).  As you can see in the contributions and the readme file, NXP has been involved, contributed, supports it, and intends to do so in the future. But, like any contributor, we cannot and will not assume responsibility or liability beyond the terms of the BSD style license.

As to JenNet-IP (http://www.nxp.com/techzones/wireless-connectivity/jennet-ip.html), this is a commercial product supported by NXP.  It has a different architecture (mesh-under) and a different target community, although overlap may exist.  As choice is good for customers, and Contiki is a well-respected Open Source option that now is also available on the JN516x family, I see no conflict but a strengthening of the offering.

**Comment by simonduq** [comment]
This PR fulfills the new platform requirements, got only positive feedback, and does not affect core nor existing platforms. As one of the long-term maintainers, I'm going to merge in a week or two if there is no objection.

**Comment by nvt** [comment]
:+1:

**Pull request merged**